[Dynamic Form] Fix for 1788/1794 and add styles property for support customizing styling

wuxiaojun514 opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | [x]
| New feature?    | [x]
| New sample?      | [ ]
| Related issues?  | fixes #1788 , partially #1794 

#### What's in this Pull Request?

This PR is mainly for 

- Dynamic Form has quite a lot of problems when [custom list formatting](https://learn.microsoft.com/en-us/sharepoint/dev/declarative-customization/list-form-configuration) is enabled. I fixed the issues (#1788  ,#1794 Bad UI) .
The author of #1794 provided some other approach to customize the styling but I feel this PR should work.
Now it will show 3 columns per row if it has custom form layout
![image](https://github.com/user-attachments/assets/55591f74-6fb4-4d29-a496-03e20af9f4a7)

-  Added two properties (styles and className) to support customization on styling and converted this component as [style-able component](https://github.com/microsoft/fluentui/blob/master/docs/react-wiki-archive/Component-Styling.md#creating-a-styleable-component).  Deleted the **/dynamicForm/DynamicForm.module.scss** file and put its content into _**/dynamicForm/DynamicForm.sytles.ts**_ and _**/dynamicForm/dynamicField/DynamicField.styles.ts**_ . I kept the same classname on styling to make this change smoothly.  


I have documented the new feature on [DynamicForm.md](https://github.com/wuxiaojun514/sp-dev-fx-controls-react/blob/fix-for-1788-and-add-styles-property/docs/documentation/docs/controls/DynamicForm.md#idynamicformstyles-interface) and given an example how to overwrite styling

I know it is quite a big change, please let me know if anything missing.




#### Guidance
- *You can delete this section when you are submitting the pull request.* 
- *Please update this PR information accordingly. We'll use this as part of our release notes in monthly communications.*
- *Please target your PR to **dev** branch.*

wuxiaojun514 commented:
Hi @joelfmrodrigues , could you look into it if you have some time?
This PR is for fixing two bugs (#1788 ,#1794 ) on Dynamic Form when list has enabled section layout and introduced **className** and **styles** property on this control so developer can customize the styling on it.
I also updated the related documentation for the new feature.
This PR changed quite a lot of code so it is very easy to have conflict with other submission ( I already fixed the conflicts two times), so I hope it can be reviewed early...
Please contact me if anything unclear or I need modify it.

Marinofull commented:
Hi @joaojmendes any updates here?

CodingSinceThe80s commented:
Hi all, just wanted to add my feedback: this would be a very useful feature so looking forward to it 👍

joelfmrodrigues commented:
@wuxiaojun514 apologies for the delay, I have seen the email notification when you tagged me a while ago but things have been a bit chaotic (in a good way 😇 ) at my end and I have been unable to look into it

wuxiaojun514 commented:
Hi @joelfmrodrigues , never mind, just take your time 😄

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
 | storeLastActiveTab | boolean | no |  When uploading files: Specifies if last active tab will be stored after the Upload panel has been closed. Note: the value of selected tab is stored in the queryString hash. Default - `true` |
 | folderPath | string | no | Server relative or library relative folder to create the item in. This option is only available for document libraries and works only when the contentTypeId is specified and has a base type of type Document or Folder. Defaults to the root folder of the library. |
+| className | string | no | Set CSS Class. |
+| styles | IStyleFunctionOrObject&lt;IDynamicFormStyleProps, [IDynamicFormStyles](#IDynamicFormStyles)&gt; | no | Styles to apply on control.  See the example [here](#how-to-use-styles-property) |
```
```suggestion
| styles | IStyleFunctionOrObject&lt;IDynamicFormStyleProps, [IDynamicFormStyles](#idynamicformstyles-interface)&gt; | no | Styles to apply on control.  See the example [here](#how-to-use-styles-property) |
```

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
 
 ## Validation Error Dialog Properties `IValidationErrorDialogProps`
-
 | Property                    | Type    | Required | Description                                                                      |
```
Please add a line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
 | customMessage               | string  | no       | Specifies a custom message to be shown in the validation dialog. Default - empty |
+
+## IDynamicFormStyles interface
+| Property | Type  | Description |
```
Please add a line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| buttons | IStyle | styles for the buttons (save button/cancel button)  |
+| subComponentStyles | {fieldStyles: [IDynamicFieldStyles](#IDynamicFieldStyles)} | styles of dynamic field control|
+
+
```
Please remove one line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+
+
+## IDynamicFieldStyles interface
+| Property | Type  | Description |
```
Please add a line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| thumbnailFieldButtons | IStyle | styles for button when field type is 'Thumbnail' |
+| selectedFileContainer | IStyle | styles for File Selection Control |
+
+
```
Please remove one line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
````diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+## How to use styles property 
+Property styles of Dynamic Form gives you a set of properties which you can use to modify styles.
+In this example it shows 4 columns (by default it shows 3 columns per row) in one row if screen size is bigger than 1280px and make the error message font size a bit large.
+```TypeScript
````
Please add a line break.

wuxiaojun514 replied:
![image](https://github.com/user-attachments/assets/7df3227b-92ab-482c-aee6-76cf53a047d1)
Hi @michaelmaillot ,

I added the line break before the sample code, is it correct?

michaelmaillot replied:
Yes!

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+
+
+## How to use styles property 
+Property styles of Dynamic Form gives you a set of properties which you can use to modify styles.
```
Please add a line break.

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| selectedFileContainer | IStyle | styles for File Selection Control |
+
+
+## How to use styles property 
```
```suggestion
## How to use styles property
```

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+            subComponentStyles:{
+              fieldStyles:{
+                errormessage:{
+                  "font-size":"18px" //overwrite the error message font size in Dynamic Field
```
I was not able to make this work. I tried it with a required Lookup field, without success:

![image](https://github.com/user-attachments/assets/7a00cf02-afb5-4f67-87a4-7302e9acd67e)

Did I miss something?

wuxiaojun514 replied:
It doesn't work  on my side as well.  Currently `subComponentStyles `is not working.     I will fix it and let you know

wuxiaojun514 replied:
@michaelmaillot I have fixed it now.
I also updated the sample in documentation https://github.com/wuxiaojun514/sp-dev-fx-controls-react/blob/fix-for-1788-and-add-styles-property/docs/documentation/docs/controls/DynamicForm.md#how-to-use-styles-property 
to handle normal text field (the error message in normal text field is generated inside the fluent ui [TextField](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield) control, so we need use `selectors `to overwrite it.

```
          styles={{
            sectionFormField: {
              selectors: {
                ':has(div)': {
                  [`@media (min-width: 1280px)`]: {
                    "min-width": '21%',
                    "max-width": '21%' //force show 4 columns per row in big screen size when enabled list custom formatting
                  }
                },
              },
            },

            subComponentStyles: {
              fieldStyles: {
                errormessage: {
                  "font-size": "18px" //overwrite the error message font size in Dynamic Field for Lookup/Note/Date fields (The error message element is generated directly in Dynamic Field control)
                },
                fieldDisplay: {
                  selectors: {
                    '.ms-TextField-errorMessage': {
                      "font-size": "18px !important" //overwrite the error message font size in Dynamic Field for TextField (error message element is wrapped in Fluent UI TextField control, we cannot modify it directly,we have to use 'selectors' to change it)
                    }
                  }
                }
              }
            }
          }}
```

michaelmaillot commented on src/controls/dynamicForm/DynamicForm.sytles.ts:
```diff
@@ -0,0 +1,152 @@
+import { IStyle } from '@fluentui/react';
```
I think this file should be named "DynamicForm.**styles**.ts" not "DynamicForm.**sytles**.ts"

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| Property | Type  | Description |
+| ---- | ---- | ---- |
+| root | IStyle | styles for the root element |
+| sectionTitle | IStyle | styles for the **section title** when your list has enabled [list formatting on form layout](https://learn.microsoft.com/en-us/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-body-with-one-or-more-sections) |
```
```suggestion
| sectionTitle | IStyle | styles for the **section title** when your list has enabled [list formatting on form layout](https://learn.microsoft.com/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-body-with-one-or-more-sections) |
```

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| sectionFormFields | IStyle | styles for the **section container** when your list has enabled list formatting on form layout |
+| sectionFormField | IStyle | styles for the **section field** when your list has enabled list formatting on form layout |
+| sectionLine | IStyle | styles for the **section line break** when your list has enabled list formatting on form layout |
+| header | IStyle | styles for the **header** when your list has enabled list formatting on [custom header](https://learn.microsoft.com/en-us/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-header) |
```
```suggestion
| header | IStyle | styles for the **header** when your list has enabled list formatting on [custom header](https://learn.microsoft.com/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-header) |
```

michaelmaillot commented on docs/documentation/docs/controls/DynamicForm.md:
```diff
@@ -69,11 +69,75 @@ The `DynamicForm` can be configured with the following properties:
+| sectionFormField | IStyle | styles for the **section field** when your list has enabled list formatting on form layout |
+| sectionLine | IStyle | styles for the **section line break** when your list has enabled list formatting on form layout |
+| header | IStyle | styles for the **header** when your list has enabled list formatting on [custom header](https://learn.microsoft.com/en-us/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-header) |
+| footer | IStyle | styles for the **footer** when your list has enabled list formatting on [custom footer](https://learn.microsoft.com/en-us/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-footer) |
```
```suggestion
| footer | IStyle | styles for the **footer** when your list has enabled list formatting on [custom footer](https://learn.microsoft.com/sharepoint/dev/declarative-customization/list-form-configuration#configure-custom-footer) |
```

michaelmaillot reviewed (changes requested):
Hey @wuxiaojun514, this is a great improvment, thanks!

I made some suggestions and one issue when trying to test on my side, please have a look and let me know.

wuxiaojun514 commented:
Hi @michaelmaillot 
Much thanks for your review. 
I appreciate your guide and testing!
Now I have updated the documentation with your advice and I have fixed typo issues in the code.
Please let me know if anything else needs fixing.

michaelmaillot commented on src/controls/dynamicForm/dynamicField/IDynamicFieldProps.ts:
```diff
@@ -92,4 +93,30 @@ export interface IDynamicFieldProps {
+ fieldLabel:IStyle;
+ labelContainer:IStyle;
+ pickersContainer:IStyle;
+ FieldEditor:IStyle;
```
Shouldn't it be written like this?

```suggestion
 fieldEditor:IStyle;
```

wuxiaojun514 replied:
I have corrected the class name and related document (Actually I copied this name from old _DynamicForm.module.scss_ 
 file...)

michaelmaillot reviewed (changes requested):
I tried again and I confirm it works now!

Just a few updates and we'll be good!

michaelmaillot commented:
Merged manually.

Thank you!
